### PR TITLE
Mark-compact GC: update STACK_TOP when growing mark stack

### PR DESCRIPTION
**Pull request by osa1**
When growing mark stack we should update `STACK_TOP` otherwise subsequent
pushes will think that the stack is full and call `grow_stack` again.

Discovered while debugging strange allocation pattern in #2650 with the
compacting GC. It turns out once the mark stack is full ever subsequent push
(until stack is smaller than the initial size again) doubles the stack size,
causing redundant allocations.

Regression test added.

### CODE REVIEW

**Review by crusso — APPROVED** [review]
Is this a bug fix?

**Comment by osa1** [comment]
@crusso

> Is this a bug fix?

Yes, updated the PR description.

File: rts/motoko-rts/src/gc/mark_compact/mark_stack.rs [hunk 1]
```diff
@@ -13,13 +13,13 @@ pub const INIT_STACK_SIZE: Words<u32> = Words(64);
 static mut STACK_BLOB_PTR: *mut Blob = null_mut();
```
Comment (crusso) [inline rts/motoko-rts/src/gc/mark_compact/mark_stack.rs:13]: Keeping that one private deliberately?

**Review by crusso — APPROVED** [review]
Still LGTM.

Does this explain the weird GC cliff Andreas observed in your earlier PR? Probably not, but might be worth checking

File: rts/motoko-rts/src/gc/mark_compact/mark_stack.rs (hunk 1 above)
Comment (osa1) [reply rts/motoko-rts/src/gc/mark_compact/mark_stack.rs:13]: It's not used in tests so no need to make it `pub`.  Should I make it `pub` for consistency?

**Comment by osa1** [comment]
> Does this explain the weird GC cliff Andreas observed in your earlier PR? Probably not, but might be worth checking

This fixes that, yes.

**Comment by crusso** [comment]
Yeah, sorry, that was actually clear from your updated PR description ;->

**Pull request merged**